gemini/sample-apps/fashion-trends-prediction/app/data_prep/driver.py

```diff
@@ -0,0 +1,332 @@
+    for country in saved["finaldata"]:
+        if country != "India":
+            for key in saved["finaldata"][country]:
+                temp = saved["finaldata"][country][key]
+                saved["finaldata"][country][key] = [[item, "", ""] for item in temp]
+
+    with open(data_path, "w") as outfile:
+        json.dump(saved, outfile)
+
+
+def create_news_articles_data():
+    """Extracts news articles from Vogue and saves them in a JSON file."""
+
+    with open(data_path, "r") as f:
+        saved = json.load(f)
+
+    if "articles" not in saved:
+        saved["articles"] = []
+
+    url = "https://www.vogue.in"
+
+    past_scrape = saved["articles"]
+    saved["articles"] = get_articles(url, past_scrape, numPages=10)
+
+    with open(data_path, "w") as outfile:
+        json.dump(saved, outfile)
+
+
+def create_article_attributes():
+    """Extracts attributes from news articles and saves them in a JSON file."""
+
+    with open(data_path, "r") as f:
+        saved = json.load(f)
+
+    print("creating article attributes")
+    saved["article_attributes"] = {}  # make new everytime
+
+    for article in saved["articles"]:
+        article_link = article[0]
+        article_summary = article[1]
+        article_attr = article[2]
+        for item in article_attr:
+            if (
+                (article_attr[item] == "none")
+                or (article_attr[item] == "None")
+                or isinstance(item, str) is False
+            ):
+                print("invalid \n  ", item, article_attr[item])
+                continue
+            if item in saved["article_attributes"]:
+                saved["article_attributes"][item].append(
+                    [article_attr[item], article_summary, article_link]
+                )
+            else:
+                saved["article_attributes"][item] = [
+                    [article_attr[item], article_summary, article_link]
+                ]
+
+    with open(data_path, "w") as outfile:
+        json.dump(saved, outfile)
+
+
+def get_top_categories(saved):
+    """Gets the top categories from the final data file.
+
+    Args:
+            saved (dict): A dictionary containing the saved data.
+
+    """
+
+    saved["top_categories"] = {}
+    for country in saved["finaldata"]:
+        allcategories = saved["finaldata"][country]
+
+        sorted_a = sorted(allcategories.items(), key=cmpf)
+        item_categories = [cat for cat, values in sorted_a[: min(15, len(sorted_a))]]
+
+        saved["top_categories"][country] = item_categories
+
+    with open(data_path, "w") as outfile:
+        json.dump(saved, outfile)
+
+
+def instaScrape():
+    """Scrapes data from Instagram and saves it in a JSON file."""
+
+    scrape_size = {"num_countries": 10, "num_influencers": 20, "num_posts": 50}
+
+    # set first to True if this is the first time you are extracting
+    periodic_extraction(scrape_size=scrape_size, first_period=True)
+
+    with open(data_path, "r") as f:
```

Comment:
Why is this code here twice?

Reply:
Because each function is updating the data at the source so we need to read it again before running the next function.